src/org/mozilla/javascript/NativeBigInt.java

```diff
@@ -0,0 +1,216 @@
+            return val;
+
+        } else if (id < Id_constructor) {
+            return execConstructorCall(id, args);
+        }
+
+        // The rest of BigInt.prototype methods require thisObj to be BigInt
+        BigInteger value = ensureType(thisObj, NativeBigInt.class, f).bigIntValue;
+
+        switch (id) {
+
+          case Id_toString:
+          case Id_toLocaleString:
+            {
+                // toLocaleString is just an alias for toString for now
+                int base = (args.length == 0 || args[0] == Undefined.instance)
+                    ? 10 : ScriptRuntime.toInt32(args[0]);
+                return ScriptRuntime.bigIntToString(value, base);
+            }
+
+          case Id_toSource:
+              return "(new BigInt("+ScriptRuntime.toString(value)+"))";
+
+          case Id_valueOf:
+              return value;
+
+          default: throw new IllegalArgumentException(String.valueOf(id));
+        }
+    }
+
+    private static Object execConstructorCall(int id, Object[] args)
+    {
+        switch (id) {
+        case ConstructorId_asIntN:
+        case ConstructorId_asUintN: {
+            int bits = ScriptRuntime.toIndex(args.length < 1 ? Undefined.instance : args[0]);
+            BigInteger bigInt = ScriptRuntime.toBigInt(args.length < 2 ? Undefined.instance : args[1]);
+
+            if (bits == 0) {
+                return BigInteger.ZERO;
+            }
+
+            byte[] bytes = bigInt.toByteArray();
+
+            int newBytesLen = (bits / Byte.SIZE) + 1;
+            if (newBytesLen > bytes.length) {
+                return bigInt;
+            }
+
+            byte[] newBytes = Arrays.copyOfRange(bytes, bytes.length - newBytesLen, bytes.length);
+
+            int mod = bits % Byte.SIZE;
+            switch (id) {
+            case ConstructorId_asIntN:
+                if (mod == 0) {
+                    newBytes[0] = newBytes[1] < 0 ? (byte)-1 : 0;
+                } else if ((newBytes[0] & (1 << (mod - 1))) != 0) {
+                    newBytes[0] |= -1 << mod;
+                } else {
+                    newBytes[0] &= (1 << mod) - 1;
+                }
+                break;
+            case ConstructorId_asUintN:
+                newBytes[0] &= (1 << mod) - 1;
+                break;
+            }
+            return new BigInteger(newBytes);
+        }
+
+        default:
+            throw new IllegalArgumentException(String.valueOf(id));
+        }
+    }
+
+    @Override
+    public String toString() {
+        return ScriptRuntime.bigIntToString(bigIntValue, 10);
+    }
+
+    @Override
+    protected int findPrototypeId(Symbol k)
+    {
+        if (SymbolKey.TO_STRING_TAG.equals(k)) {
+            return SymbolId_toStringTag;
+        }
+        return 0;
+    }
+
+// #string_id_map#
+
+    @Override
+    protected int findPrototypeId(String s)
+    {
+        int id;
+// #generated# Last update: 2007-05-09 08:15:50 EDT
```

Comment:
https://github.com/mozilla/rhino/blob/3e5eea461b7500f3c25793553f76c6a831b7f372/src/org/mozilla/javascript/NativeBigInt.java#L179

Is this really fast? I feel it would be easier for the JVM to optimize using Strings in switch in Java7.
https://docs.oracle.com/javase/8/docs/technotes/guides/language/strings-switch.html

Reply:
ref. #847